app/src/file_handler.c

```diff
@@ -0,0 +1,195 @@
+        return SDL_FALSE;
+    }
+    // transfer ownership
+    *file = queue->data[queue->tail];
+    queue->tail = (queue->tail + 1) % FILE_QUEUE_SIZE;
+    return SDL_TRUE;
+}
+
+SDL_bool file_handler_init(struct file_handler *file_handler, const char *serial) {
+
+    if (!file_queue_init(&file_handler->queue)) {
+        return SDL_FALSE;
+    }
+
+    if (!(file_handler->mutex = SDL_CreateMutex())) {
+        return SDL_FALSE;
+    }
+
+    if (!(file_handler->event_cond = SDL_CreateCond())) {
+        SDL_DestroyMutex(file_handler->mutex);
+        return SDL_FALSE;
+    }
+
+    if (serial) {
+        file_handler->serial = SDL_strdup(serial);
+        if (!file_handler->serial) {
+            LOGW("Cannot strdup serial");
+            SDL_DestroyMutex(file_handler->mutex);
+            return SDL_FALSE;
+        }
+    } else {
+        file_handler->serial = NULL;
+    }
+
+    // lazy initialization
+    file_handler->initialized = SDL_FALSE;
+
+    file_handler->stopped = SDL_FALSE;
+    file_handler->current_process = PROCESS_NONE;
+
+    return SDL_TRUE;
+}
+
+void file_handler_destroy(struct file_handler *file_handler) {
+    SDL_DestroyCond(file_handler->event_cond);
+    SDL_DestroyMutex(file_handler->mutex);
+    file_queue_destroy(&file_handler->queue);
+    SDL_free((void *) file_handler->serial);
+}
+
+SDL_bool file_handler_do(struct file_handler *file_handler, const char *file) {
+    SDL_bool res;
+
+    // start file_handler if it's used for the first time
+    if (!file_handler->initialized) {
+        if (!file_handler_start(file_handler)) {
+            return SDL_FALSE;
+        }
+        file_handler->initialized = SDL_TRUE;
+    }
+
+    mutex_lock(file_handler->mutex);
+    SDL_bool was_empty = file_queue_is_empty(&file_handler->queue);
+    res = file_queue_push(&file_handler->queue, file);
+    if (was_empty) {
+        cond_signal(file_handler->event_cond);
+    }
+    mutex_unlock(file_handler->mutex);
+    return res;
+}
+
+static int is_apk(const char *file) {
+    const char *ext = strrchr(file, '.');
+    return ext && !strcmp(ext, ".apk");
+}
+
+static int run_file_handler(void *data) {
+    struct file_handler *file_handler = data;
+
+    for (;;) {
+        mutex_lock(file_handler->mutex);
+        while (!file_handler->stopped && file_queue_is_empty(&file_handler->queue)) {
+            cond_wait(file_handler->event_cond, file_handler->mutex);
+        }
+        if (file_handler->stopped) {
+            // stop immediately, do not process further events
+            mutex_unlock(file_handler->mutex);
+            break;
+        }
+        char *current_file;
+        char cmd[16];
+        process_t process;
+#ifdef BUILD_DEBUG
+        bool non_empty = file_queue_take(&file_handler->queue, &current_file);
+        SDL_assert(non_empty);
+#else
+        file_queue_take(&file_handler->queue, &current_file);
+#endif
+        LOGI("Processing %s...", current_file);
+        if (is_apk(current_file)) {
```

Comment:
Hi, @rom1v 

Give a type to every request is a good idea!
However, I think that push an apk without installing it make no sense. When do we need to just push an apk on the device and install it later?

Thank you for your time and effort.

Reply:
It seems weird to be able to push any file except a specific type. One may want to use its device as an "usb key", and store apk to copy them somewhere afterwards.

It's not very important though.

Reply:
Hi, @rom1v 

Push an apk without installing it is an rare case. I don't think that break a good design because of unusual situations is a good idea. We should remain intuitive and convenient method (simply drag & drop) for user. Therefore, I propose these behavior:

1. Install apk (common case, just drag & drop)
2. Push file (common case, just drag & drop)
3. Push apk (rare case, ctrl + drag & drop)

What do you think?

Regards!

Reply:
OK, so for simplicity, just forget 3. :wink: